src/ATokenNaive.sol

```diff
@@ -0,0 +1,357 @@
+  function castVote(uint256 proposalId) external {
+    if (internalVotingPeriodEnd(proposalId) > block.number) revert("cannot castVote yet");
+    uint8 unusedSupportParam = uint8(VoteType.Abstain);
+    ProposalVote memory _proposalVote = proposalVotes[proposalId];
+
+    uint256 _proposalSnapshotBlockNumber = governor.proposalSnapshot(proposalId);
+
+    // Use the snapshot of total deposits to determine total voting weight. We cannot
+    // use the proposalVote numbers alone, since some people with deposits at the
+    // snapshot might not have expressed votes.
+    uint256 _totalDepositWeightAtSnapshot = getPastTotalDeposits(_proposalSnapshotBlockNumber);
+
+    // We need 256 bits because of the multiplication we're about to do.
+    uint256 _votingWeightAtSnapshot = IVotingToken(address(_underlyingAsset)).getPastVotes(
+      address(this), _proposalSnapshotBlockNumber
+    );
+
+    //      forVotesRaw          forVotesScaled
+    // --------------------- = ---------------------
+    //     totalDeposits        deposits (@snapshot)
+    //
+    // forVotesScaled = forVotesRaw * deposits@snapshot / totalDeposits
+    uint128 _forVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.forVotes) / _totalDepositWeightAtSnapshot
+    );
+    uint128 _againstVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.againstVotes) / _totalDepositWeightAtSnapshot
+    );
+    uint128 _abstainVotesToCast = SafeCast.toUint128(
+      (_votingWeightAtSnapshot * _proposalVote.abstainVotes) / _totalDepositWeightAtSnapshot
+    );
+
+    bytes memory fractionalizedVotes =
+      abi.encodePacked(_forVotesToCast, _againstVotesToCast, _abstainVotesToCast);
+    governor.castVoteWithReasonAndParams(
+      proposalId, unusedSupportParam, "crowd-sourced vote", fractionalizedVotes
+    );
+  }
+
+  // forgefmt: disable-start
+  //===========================================================================
+  // BEGIN: Aave overrides
+  //===========================================================================
+  /// Note: this has been modified from Aave v3's AToken to call our custom
+  /// mintScaled function.
+  ///
+  /// @inheritdoc IAToken
+  function mint(
+    address caller,
+    address onBehalfOf,
+    uint256 amount,
+    uint256 index
+  ) external virtual override onlyPool returns (bool) {
+    return __mintScaled(caller, onBehalfOf, amount, index);
+  }
+
+  /// Note: this has been modified from Aave v3's AToken to call our custom
+  /// mintScaled function.
+  ///
+  /// @inheritdoc IAToken
+  function mintToTreasury(uint256 amount, uint256 index) external override onlyPool {
+    if (amount == 0) {
+      return;
+    }
+    __mintScaled(address(POOL), _treasury, amount, index);
+  }
+
+  /// Note: this has been modified from Aave v3's AToken to update deposit
+  /// balance accordingly. We cannot just call `super` here because the function
+  /// is external.
```

Comment:
Huh TIL. Okay, if you prefer this I'll just leave as-is

Reply:
Sorry, to clarify I think we should call `super.burn` or `AToken.burn` instead of copy/pasting the code

Reply:
Oh yea, you're right. Clearly I messed up when testing 😅

We can leave as-is for this PR but I think we should lean towards upstream PRs to Aave's codebase so we can avoid copying code as much as possible. For example PR in a `public virtual` change here instead of `external`

Reply:
@davidlaprade should we have an issue / track somewhere "changes to upstream to AToken", and include this, before marking this resolved? Feel free to resolve this comment yourself once tracking (assuming you agree we should track somewhere)

Reply:
Tracking here https://github.com/ScopeLift/flexible-voting/issues/16